default/python/AI/ProductionAI.py

```diff
@@ -1611,59 +1272,57 @@ def already_has_completed_colony_building(planet_id) -> bool:
     return any(universe.getBuilding(bldg).name.startswith("BLD_COL_") for bldg in planet.buildingIDs)
 
 
-def build_ship_facilities(bld_name, top_locs=None):
+def _build_ship_facilities(building_type: BuildingType, top_pids: Set[PlanetId] = None) -> None:
     # TODO: add total_pp checks below, so don't overload queue
-    if top_locs is None:
-        top_locs = set()
+    if top_pids is None:
+        top_pids = set()
     universe = fo.getUniverse()
     empire = fo.getEmpire()
     total_pp = empire.productionPoints
-    __, prereq_bldg, this_cost, time = AIDependencies.SHIP_FACILITIES.get(bld_name, (None, "", -1, -1))
-    if not get_aistate().character.may_build_building(bld_name):
-        return
-    bld_type = fo.getBuildingType(bld_name)
-    if not empire.buildingTypeAvailable(bld_name):
+    prerequisite_type = building_type.prerequisite()
```

Comment:
It's a good practice to declare a variable as close to its use as possible.  It makes code easier to follow.